### ISSUE TITLE
Feature Request : Support non-deterministic ordering of slice args in `Mock.On(...)`

### ISSUE THREAD

**Issue by manan-sc** [stretchr/testify#1117]
Assume that a function that we're mocking accepts a `slice[]` of `strings`, and it returns a response `friendData` thereafter.

A typical mock for this would be something like below - 

`serviceMock.On("GetFriends", friendIDs).Return(friendData, err)`

where, `friendIDs = []string{"abc", "def", "xyz"}`

The mock will work if (and only if)`GetFriends` is definitely called with `friendIDs = []string{"abc", "def", "xyz"}`. However, if it's called with say, `friendIDs = []string{"xyz", "abc", "def"}`, the mock will `panic` with `mock: Unexpected Method Call`.

Something like below - 

```
panic: 

mock: Unexpected Method Call
-----------------------------

GetFriends([]string)
                0: []string{"xyz", "abc", "def"}

The closest call I have is: 

GetFriends([]string)
                0: []string{"abc", "def", "xyz"}

Difference found in argument 0:

--- Expected
+++ Actual
@@ -1,4 +1,4 @@
 ([]string) (len=2) {
- (string) (len=3) "abc",
- (string) (len=3) "def",
- (string) (len=3) "xyz",
- (string) (len=3) "xyz",
- (string) (len=3) "abc",
- (string) (len=3) "def",
 }

Diff: 
        0: FAIL:  ([]string{"xyz", "abc", "def"}) != ([]string{"abc", "def", "xyz"}) [recovered]
        panic: 

```

It could be that the `friendIDs` are returned by another service (or a DB). Using `mock.Anything` (although it works) seems a bit _misleading_ - it's not as if we don't care as to what argument was used to call the function, just that the ordering of individual elements inside that slice is immaterial.

Is there an intention to support `Any` ordering of elements inside a `slice`, so long as all the elements (and only those elements) are there present in the argument?

**Comment by brackendawson** [comment]
I don't think any built in argument matcher for slices in any order is likely to be added, maybe after generics.

However you can build your own argument matchers using `mock.MatchedBy()`, here's one using the well known elements match algorithm:
```go
package kata

import (
	"testing"

	"github.com/stretchr/testify/mock"
)

func TestCompare(t *testing.T) {
	m := &myMock{}
	m.Test(t)
	m.On("MyFunc", mock.MatchedBy(func(friends []string) bool {
		return elementsMatch(friends, []string{"barney", "fred", "wilma"})
	})).Return(nil)
	defer m.AssertExpectations(t)

	m.MyFunc([]string{"fred", "wilma", "barney"})
}

func elementsMatch(x, y []string) bool {
	if len(x) != len(y) {
		return false
	}
	diff := make(map[string]int, len(x))
	for _, _x := range x {
		diff[_x]++
	}
	for _, _y := range y {
		if _, ok := diff[_y]; !ok {
			return false
		}
		diff[_y] -= 1
		if diff[_y] == 0 {
			delete(diff, _y)
		}
	}
	return len(diff) == 0
}

type myMock struct {
	mock.Mock
}

func (m *myMock) MyFunc(friends []string) error {
	return m.Called(friends).Error(0)
}
```

**Comment by manan-sc** [comment]
> I don't think any built in argument matcher for slices in any order is likely to be added, maybe after generics.
> 
> However you can build your own argument matchers using `mock.MatchedBy()`, here's one using the well known elements match algorithm:
> 
> ```go
> package kata
> 
> import (
> 	"testing"
> 
> 	"github.com/stretchr/testify/mock"
> )
> 
> func TestCompare(t *testing.T) {
> 	m := &myMock{}
> 	m.Test(t)
> 	m.On("MyFunc", mock.MatchedBy(func(friends []string) bool {
> 		return elementsMatch(friends, []string{"barney", "fred", "wilma"})
> 	})).Return(nil)
> 	defer m.AssertExpectations(t)
> 
> 	m.MyFunc([]string{"fred", "wilma", "barney"})
> }
> 
> func elementsMatch(x, y []string) bool {
> 	if len(x) != len(y) {
> 		return false
> 	}
> 	diff := make(map[string]int, len(x))
> 	for _, _x := range x {
> 		diff[_x]++
> 	}
> 	for _, _y := range y {
> 		if _, ok := diff[_y]; !ok {
> 			return false
> 		}
> 		diff[_y] -= 1
> 		if diff[_y] == 0 {
> 			delete(diff, _y)
> 		}
> 	}
> 	return len(diff) == 0
> }
> 
> type myMock struct {
> 	mock.Mock
> }
> 
> func (m *myMock) MyFunc(friends []string) error {
> 	return m.Called(friends).Error(0)
> }
> ```

Nice, thanks @brackendawson - that should certainly work! I'd like to know why you think that this feature will be more useful until after there's support for generics in go.

**Comment by brackendawson** [comment]
> I'd like to know why you think that this feature will be more useful until after there's support for generics in go.

I'm probably just being a bit thick, adding any generic code to the repo would mean 1.18 being the minimum supported Go version. Thinking some more we could definitely just use reflection and have a function in mock like:
```go
// MatchedByElements can be used to match a mock call based the elements of a
// slice or array ignoring the order of the elements.
func MatchedByElements(list interface{}) argumentMatcher
```
(argumentMatcher might not be the best return value but it could certainly work).

Then you can just:
```go
mockedThing.On("FunctionName", mock.MatchedByElements([]string{"tom", "dick", "harry"}).Return("things")
```
I'd ✅  that, I wonder what the maintainers would think?

**Comment by manan-sc** [comment]
> > I'd like to know why you think that this feature will be more useful until after there's support for generics in go.
> 
> I'm probably just being a bit thick, adding any generic code to the repo would mean 1.18 being the minimum supported Go version. Thinking some more we could definitely just use reflection and have a function in mock like:
> 
> ```go
> // AnyOrder can be used to match a mock call based the elements of a
> // slice or array ignoring the order of the elements.
> func AnyOrder(list interface{}) argumentMatcher
> ```
> 
> (argumentMatcher might not be the best return value but it could certainly work).
> 
> Then you can just:
> 
> ```go
> mockedThing.On("FunctionName", mock.AnyOrder([]string{"tom", "dick", "harry"}).Return("things")
> ```
> 
> I'd ✅ that, I wonder what the maintainers would think?

Yeah, that sounds good to me - the semantics around `AnyOrder` indicating that the argument is to be matched by _something_ that has the same elements (with their individual frequencies), and is agnostic of the relative element ordering.

**Comment by yusrilmr** [comment]
I am fond of @brackendawson 's idea with `mock.AnyOrder()`.
I hope the maintainers will consider it.

**Comment by amitavaghosh1** [comment]
Is it possible to atleast expose the mock.argumentMatcher , so that we can do. 
It's quite clever, how you completely hid the implementation. by making `On` receive `interface{}`. But we would atleast like to do.

```go
func mockAnyOrder[E comparable](expects []E) interface{} {
	return mock.MatchedBy(func(vals []E) bool {
		return hasDifference(vals, expects)
	})
}

func hasDifference[E comparable](a, b []E) bool {
	mb := make(map[E]struct{}, len(b))
	for _, x := range b {
		mb[x] = struct{}{}
	}

	for _, x := range a {
		if _, found := mb[x]; !found {
			return true
		}
	}

	return false
}
```

I tried this but started getting errors like.

```txt
FindBy(*context.emptyCtx,[]string)
                0: (*context.emptyCtx)(0xc00001c0b8)
                1: []string{"nc1", "uc1", "ec1"}

The closest call I have is:

FindBy(*context.emptyCtx,mock.argumentMatcher)
                0: (*context.emptyCtx)(0xc00001c0b8)
                1: mock.argumentMatcher{fn:reflect.Value{typ:(*reflect.rtype)(0x13183e0), ptr:(unsafe.Pointer)(0xc000422120), flag:0x13}}
```

**Comment by botkevin** [comment]
I hope the maintainers will add this method or at least make

**Comment by manan-sc** [comment]
> 

Well, you could use something like below (taking inspiration from @brackendawson's idea) - 

```go

// Test body in case of general bool-type assertions
assert.True(t, ElementsMatch([]string{a, b}, []string{b, a}))

// Test body in case of fn argument matcher assertions
userIDs := []string{a, b}
funcMock.On("GetFoo", mock.MatchedBy(func(someIDs []string) bool {
		return ElementsMatch(someIDs, userIDs)
	})).Return(fooData, err)


// ElementsMatch asserts that the specified listA(array, slice...) is equal to specified
// listB(array, slice...) ignoring the order of the elements. If there are duplicate elements,
// the number of appearances of each of them in both lists should match.
//
// ElementsMatch([1, 3, 2, 3], [1, 3, 3, 2])
func ElementsMatch(listA, listB []string) bool {
	if len(listA) != len(listB) {
		return false
	}
	diff := make(map[string]int, len(listA))
	for _, a := range listA {
		diff[a]++
	}

	for _, b := range listB {
		if _, ok := diff[b]; !ok {
			return false
		}
		diff[b]--
		if diff[b] == 0 {
			delete(diff, b)
		}
	}
	return len(diff) == 0
}
```

**Comment by botkevin** [comment]
Yes but it gets really verbose when you have multiple mocks. It would be nice to not have to repeat the code.

**Comment by manan-sc** [comment]
> Yes but it gets really verbose when you have multiple mocks. It would be nice to not have to repeat the code, and it would improve readability.

You needn't repeat the code if you put the `ElementsMatch` function in a helper package, export it, and use it wherever you want it. Perhaps that's what you're after, if I'm understanding this correctly?

**Comment by brackendawson** [comment]
Unless we put something in Testify then any hacks we cause you to repeat yourself, I'll open a PR for my idea so it's ready next time a maintainer does a drive-by, we see them from time to time.

Here's an [especially bad way](https://go.dev/play/p/CK-TFcD4Akj) of hacking it that repeats a little less code:
```go
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/mock"
)

type myMock struct {
	mock.Mock
}

func (m *myMock) Process(elems []string) error {
	return m.Called(elems).Error(0)
}

func TestIfy(t *testing.T) {
	m := &myMock{}
	m.Test(t)
	m.AssertExpectations(t)
	m.On("Process", mock.MatchedBy(AnyOrder([]string{"a", "b", "c"}))).Return(nil).Once()
	assert.NoError(t, m.Process([]string{"c", "a", "b"}))
}

func AnyOrder[T any](expectation []T) func([]T) bool {
	return func(elems []T) bool {
		return assert.ElementsMatch(&testing.T{}, elems, expectation)
	}
}
```

**Comment by botkevin** [comment]
No. I mean repeating 
```
funcMock.On("GetFoo", mock.MatchedBy(func(someIDs []string) bool {
		return ElementsMatch(someIDs, userIDs)
	})).Return(fooData, err)
```
instead of a more simple
```
funcMock.On("GetFoo", MockAnyOrder{userIDs}).Return(fooData, err)
```
```
